<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
 "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <link rel="stylesheet" type="text/css" media="screen, projection, print"
    href="http://www.w3.org/Talks/Tools/Slidy2/styles/slidy.css" />
  <script src="http://www.w3.org/Talks/Tools/Slidy2/scripts/slidy.js.gz"
    charset="utf-8" type="text/javascript"></script>
</head>
<body>
<div class="section slide level2" id="overview">
<h1 id="overview">Overview</h1>
<p>This document describes the cmph5tools.py query interface. This interface is used to both produce data tables as well as new cmp.h5 files. The interface is meant to be somewhat similar to SQL.</p>
<p>At the heart of the new tools is a small query language for extracting alignments and computing statistics over those alignments. The three relevant clauses are: <code>what</code>, <code>where</code>, and <code>groupBy</code>.</p>
</div>
<div class="section slide level2" id="example-1-produce-a-sub-sampled-cmp.h5-file">
<h1 id="example-1-produce-a-sub-sampled-cmp.h5-file">Example 1: produce a sub-sampled cmp.h5 file</h1>
<pre><code>$ export INCMP=$HOME/projects/software/bioinformatics/tools/pbh5tools/etc/aligned_reads_ss.cmp.h5</code></pre>
<p>take 50% of the reads</p>
<pre><code>$ cmph5tools.py select --where &quot;SubSample(.5)&quot; $INCMP --outFile ss.cmp.h5
$ test -f ss.cmp.h5</code></pre>
</div>
<div class="section slide level2" id="example-2-produce-tabular-data-from-a-cmp.h5-file">
<h1 id="example-2-produce-tabular-data-from-a-cmp.h5-file">Example 2: produce tabular data from a cmp.h5 file</h1>
<pre><code>$ cmph5tools.py stats --what &quot;Tbl(q = Percentile(ReadLength, 90), m = Median(Accuracy))&quot; \
&gt; --groupBy Barcode $INCMP | tail
bc_88--bc_88      486.40          0.91
bc_89--bc_89      561.00          0.91
bc_9--bc_9        479.80          0.90
bc_90--bc_90      563.60          0.89
bc_91--bc_91      554.60          0.91
bc_92--bc_92      523.00          0.90
bc_93--bc_93      542.00          0.90
bc_94--bc_94      518.00          0.90
bc_95--bc_95      512.20          0.91
bc_96--bc_96      609.60          0.92</code></pre>
</div>
<div class="section slide level2" id="example-3-query-the-package-to-determine-the-available-metrics-and-statistics">
<h1 id="example-3-query-the-package-to-determine-the-available-metrics-and-statistics">Example 3: query the package to determine the available metrics and statistics</h1>
<pre><code>$ cmph5tools.py metrics
\t\t--- Metrics --- (esc)
TemplateSpan
\tThe number of template bases covered by the read (esc)
ReadLength
NErrors
ReadDuration
FrameRate
IPD
PulseWidth
Movie
Reference
RefIdentifier
HoleNumber
ReadStart
ReadEnd
TemplateStart
TemplateEnd
MoleculeIdx
Strand
AlignmentIdx
Barcode
SubSample[rate]
\tboolean vector with true occuring at rate rate (esc)

\t\t--- Statistics --- (esc)
Mean
Median
Count
Percentile[ptile]
Round[digits]</code></pre>
</div>
<div class="section slide level2" id="example-4-familiar-sql-like-syntax">
<h1 id="example-4-familiar-sql-like-syntax">Example 4: Familiar SQL-like syntax</h1>
<pre><code>$ cmph5tools.py stats --what &quot;Tbl(a=Accuracy,b=Barcode)&quot; \
&gt; --where &quot;Barcode == &#39;bc_78--bc_78&#39;&quot; \
&gt; --groupBy Reference $INCMP
Group                a          b
MET_600_t2_2      0.96          bc_78--bc_78          
MET_600_t2_2      0.82          bc_78--bc_78          
MET_600_t2_2      0.85          bc_78--bc_78          
MET_600_t2_2      0.89          bc_78--bc_78          
MET_600_t2_2      0.87          bc_78--bc_78          
MET_600_t2_2      0.90          bc_78--bc_78          
MET_600_t2_2      0.90          bc_78--bc_78          
MET_600_t2_2      0.94          bc_78--bc_78          </code></pre>
</div>
<div class="section slide level2" id="example-5-familiar-sql-like-functions">
<h1 id="example-5-familiar-sql-like-functions">Example 5: Familiar SQL-like functions</h1>
<pre><code>$ cmph5tools.py stats --what &quot;Count(Reference)&quot; \
&gt; --where &quot;Barcode == &#39;bc_78--bc_78&#39;&quot; \
&gt; --groupBy Reference $INCMP
Group             Count(Reference)
MET_600_t2_2                     8</code></pre>
</div>
<div class="section slide level2" id="handling-ipd-pulsewidth-and-base-level-metrics">
<h1 id="handling-ipd-pulsewidth-and-base-level-metrics">Handling IPD, PulseWidth, and Base-level Metrics</h1>
<p>A constant <em>hurdle</em> with cmp.h5 is dealing with the different sized data, i.e., base-level data and alignment-level data. The stats tool provides convenience functions for dealing with this</p>
<pre><code>$ cmph5tools.py stats --where &quot;(Barcode == &#39;bc_78--bc_78&#39;) &amp; (Accuracy &gt; .95)&quot; \
&gt; --what &quot;Tbl(idx = AlignmentIdx, ipd = Median(IPD), pw = Median(PulseWidth))&quot; \
&gt; $INCMP
           ipd             pw           idx
          9.00          12.00          1683</code></pre>
<p><strong>NOTE:</strong> The ‘(’ surrounding the clauses in the where predicate are</p>
</div>
<div class="section slide level2" id="selection">
<h1 id="selection">Selection</h1>
<p>Some of our internal tools work on an entire cmp.h5 file. In particular, a good deal of the Milhouse plotting code treats an input cmp.h5 file as a single condition. This <em>pushes</em> the logic of plotting down into the individual plotting functions, e.g., <code>gAccuracyByReferenceSequence</code>. Instead, we can use tools like selection to produce plots on very arbitrary splits.</p>
<pre><code>$ cmph5tools.py select --where &quot;Accuracy &gt; .95&quot; $INCMP
$ cmph5tools.py summarize out.cmp.h5
----------------------------------------
filename: out.cmp.h5
version:  1.3.1.PB
n reads:  91
n refs:   40
n movies: 4
n bases:  15315
avg rl:   168
avg acc:  0.9658</code></pre>
</div>
<div class="section slide level2" id="selection-1">
<h1 id="selection-1">Selection</h1>
<pre><code>$ cmph5tools.py select --groupBy Barcode \
&gt; --where &quot;(Barcode == &#39;bc_42--bc_42&#39;) | (Barcode == &#39;bc_28--bc_28&#39;)&quot; $INCMP
$ cmph5tools.py merge --outFile merged.cmp.h5 bc_42--bc_42.cmp.h5 bc_28--bc_28.cmp.h5
$ cmph5tools.py stats --what &quot;Count(Reference)&quot; --groupBy Barcode merged.cmp.h5
Group             Count(Reference)
bc_28--bc_28                     6
bc_42--bc_42                    65</code></pre>
<p><strong>NOTE:</strong> One of the reasons that the cmp.h5 file format is appealing is that you don’t need to split it to access particular strata. In practice, we probably don’t want to be splitting files up; A <em>View</em>-type design pattern might be more appropriate.</p>
</div>
<div class="section slide level2" id="implementation">
<h1 id="implementation">Implementation</h1>
<p>The <em>Stats</em> interface is defined <code>pbtools.pbh5tools.Metrics</code>. Implementing new metrics and statistics is fairly straightforward. In fact, all of the heavy lifting is performed by D. Alexander’s <code>CmpH5Reader</code>.</p>
</div>
</body>
</html>
